### DISCUSSION TITLE
fwupd signature validation

### DISCUSSION THREAD

**Discussion by mschmitt** [fwupd/fwupd#7668] [Q&A] [answered]
Hey everyone, 

I've been asked to demonstrate how _fwupd_ downloads are signed and validated. 

There's a _jcat_ manifest in _/var/lib/fwupd/metadata/lvfs/metadata.xml.xz.jcat_, which references a file _firmware.xml.xz_, that's nowhere on the system. However, the checksums in the manifest suggest that _metadata.xml.xz_, in the same directory **might be** this missing _firmware.xml.xz_, and in fact, it is:

```
# jcat-tool add-alias metadata.xml.xz.jcat firmware.xml.xz metadata.xml.xz
# jcat-tool --public-keys /etc/pki/fwupd verify metadata.xml.xz.jcat 
firmware-07438-stable.xml.xz:
    PASSED sha1: OK
    PASSED sha256: OK
    PASSED pkcs7: O=Linux Vendor Firmware Project,CN=LVFS CA
    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
```

What could be the reason for the actual filename being absent in the manifest?

This is on Debian 12:

```
# fwupdmgr --version
compile   org.freedesktop.fwupd         1.8.12
compile   com.hughsie.libxmlb           0.3.10
compile   com.hughsie.libjcat           0.1.9
runtime   org.freedesktop.fwupd-efi     1.4
compile   org.freedesktop.gusb          0.3.10
runtime   com.dell.libsmbios            2.4
runtime   org.freedesktop.gusb          0.3.10
runtime   org.freedesktop.fwupd         1.8.12
runtime   org.kernel                    6.1.0-23-amd64
```

**Comment by hughsie** [comment]
> There's a jcat manifest in /var/lib/fwupd/metadata/lvfs/metadata.xml.xz.jcat, which references a file firmware.xml.xz

The firmware.xml.xz is an alias -- i.e. what the file *might* be called. e.g. see:

    jcat-tool info /var/lib/fwupd/metadata/lvfs/metadata.xml.zst.jcat
    JcatFile:
      Version:               0.1
      JcatItem:
        ID:                  firmware-07437-stable.xml.zst
        AliasId:             firmware.xml.zst

i.e. fwupd downloads metadata.xml.zst.jcat find the firmware-07437-stable.xml.zst filename, and then saves it as firmware.xml.zst in the cache. The reason we do this is that the cdn might give us a metadata.xml.zst.jcat from yesterday, in which case we want the exact metadata from yesterday -- rather than one 12 h old and one 6 hours old.

**Comment by mschmitt** [comment]
Thanks for the reply! If both the XML and the JCAT file get routinely renamed, wouldn't it make sense to have the other filename embedded in the JCAT as well?

Also, another question: How do _/etc/pki/fwupd/_ and _/var/lib/fwupd/gnupg/_ relate to each other? Is _/var/lib/fwupd/gnupg/_ only updated on the fly from _/etc/pki/fwupd/_ or may it at some later point update keys from a metadata server on the fly?

**Reply by hughsie** [reply]
> wouldn't it make sense to have the other filename embedded in the JCAT as well

It is, isn't it? Also, if you want to poke a bit, rename the jcat archive to `metadata.json.gz` and then to `gzip -d metadata.json.gz` - there are no secrets there :)

> How do /etc/pki/fwupd/

This is the location for the firmware and metadata public keys.

> and /var/lib/fwupd/gnupg/

This is the GnuPG "keyring" that the public keys get imported into. GPG is old, arcane, and I'd love to switch to pkcs7/ed25519 rather than pkcs7/gpg longer term.

**Reply by mschmitt** [reply]
> > wouldn't it make sense to have the other filename embedded in the JCAT as well
> 
> It is, isn't it? 

It actually isn't, at least not in 1.8.12 on Debian 12, that's why I added that initial `add-alias` in order to produce a somewhat encouraging "success" message on my screen.

**Reply by hughsie** [reply]
Ew, Debian ships 0.1.9 -- I think you need 0.1.12 for the alias stuff to work correctly.

**Reply by superm1** [reply]
> Ew, Debian ships 0.1.9 -- I think you need 0.1.12 for the alias stuff to work correctly.

Stable effectively isn't updated unless some cve pops up as far as I know. Testing has 0.2.0.

**Reply by mschmitt** [reply]
Aliases do seem to work, but the alias simply isn't there. 

```
# jcat-tool --public-keys /etc/pki/fwupd verify metadata.xml.xz.jcat
firmware-07440-stable.xml.xz:
    FAILED: Could not find ./firmware-07440-stable.xml.xz or ./firmware.xml.xz
Validation failed
# jcat-tool add-alias metadata.xml.xz.jcat firmware.xml.xz metadata.xml.xz
# jcat-tool --public-keys /etc/pki/fwupd verify metadata.xml.xz.jcat
firmware-07440-stable.xml.xz:
    PASSED sha1: OK
    PASSED sha256: OK
    PASSED pkcs7: O=Linux Vendor Firmware Project,CN=LVFS CA
    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
```

**Reply by mschmitt** [reply]
So I installed the snap, and packaged with it still is the same unupdated version of libjcat:

```
# fwupdmgr --version
compile   com.hughsie.libxmlb           0.3.18
compile   io.snapcraft.fwupd            6305
compile   com.hughsie.libjcat           0.1.9
compile   org.freedesktop.fwupd         1.9.23
runtime   org.freedesktop.fwupd-efi     1.4
compile   org.freedesktop.gusb          0.3.10
runtime   com.hughsie.libxmlb           0.3.x
runtime   org.freedesktop.gusb          0.3.10
runtime   org.freedesktop.fwupd         1.9.23
runtime   org.kernel                    6.1.0-23-amd64
```

Is there any way to see a validating signature out of the box, without having to manipulate the _JCAT_ manifest? What makes you so confident the alias is in the _JCAT_ when in fact it isn't? Am I using this wrong?

**Reply by hughsie** [reply]
> but the alias simply isn't there

	$ wget https://cdn.fwupd.org/downloads/firmware.xml.zst.jcat
	$ jcat-tool info firmware.xml.zst.jcat 
	JcatFile:
	  Version:               0.1
	  JcatItem:
	    ID:                  firmware-07440-stable.xml.zst
	    AliasId:             firmware.xml.zst
	...

You need a new libjcat. Maybe @superm1 can update it in the snap?

**Reply by superm1** [reply]
It's because we don't have a bigger version in our subproject for 1_9_X.  This will bump it.
https://github.com/fwupd/fwupd/pull/7683

**Reply by superm1** [reply]
OK, confirmed the 1_9_X snap build pulled it in:
https://github.com/fwupd/fwupd/actions/runs/10669985348/job/29573036979

You can switch to the 1.9.x/edge build (1.9.24-4-g4a8c618a6) to pick it up.

**Comment by mschmitt** [comment]
Now validate /var/lib/fwupd/metadata/lvfs/metadata.xml.zstd using this manifest.

**Reply by hughsie** [reply]
Like this?

	wget https://cdn.fwupd.org/downloads/firmware-07440-stable.xml.zst
	jcat-tool --public-keys /etc/pki/fwupd verify firmware.xml.zst.jcat
	firmware-07440-stable.xml.zst:
	    PASSED sha1: OK
	    PASSED sha256: OK
	    PASSED pkcs7: O=Linux Vendor Firmware Project,CN=LVFS CA
	    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
	$ mv firmware-07440-stable.xml.zst firmware.xml.zst
	$ jcat-tool --public-keys /etc/pki/fwupd verify firmware.xml.zst.jcat
	firmware-07440-stable.xml.zst:
	    PASSED sha1: OK
	    PASSED sha256: OK
	    PASSED pkcs7: O=Linux Vendor Firmware Project,CN=LVFS CA
	    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2

Is the issue that fwupd saves the file as `metadata.xml.zst` rather than `firmware.xml.zst`? I though `verify` would accept a second optional argument of the source, but that appears to not be working as I remember.

**Reply by mschmitt** [reply]
This is what I've been trying to convey since my first post. Why the dance with an own signature container format that allows filename aliases when the feature is then subverted by liberally renaming everything and not having an alias for it?

**Reply by hughsie** [reply]
I'm not sure if you're filing a bug or just letting off steam; either way please be nice to us.

**Reply by mschmitt** [reply]
Sorry, english isn't my native language. Are you suggesting I should file a bug for the missing signature validation capability?

**Reply by hughsie** [reply]
@superm1 i think the easiest thing to do would be to save the metadata `/var/lib/fwupd/metadata/lvfs/firmware.xml.zst` rather than `/var/lib/fwupd/metadata/lvfs/metadata.xml.zst` -- then we can validate easier. Concur?

**Reply by superm1** [reply]
As long as we keep it in the metadata directory I think that's totally fine

**Reply by mschmitt** [reply]
Thanks to the JCAT clarifications in this topic, my corporate client here decided to permit the use of _fwupd_ on a larger number of laptops. Thanks for your work and support. :-)

**Reply by hughsie** [reply, marked as answer]
How about https://github.com/fwupd/fwupd/pull/7721 as a nice ending. :)
